src/plugins/s390.c

```diff
@@ -398,3 +399,353 @@ gchar* bd_s390_zfcp_sanitize_lun_input (const gchar *lun, GError **error) {
+                g_free (fcphbasysfs);
+                g_free (fcpwwpnsysfs);
+                g_free (fcplunsysfs);
+                g_free (fcpsysfs);
+                return FALSE;
+            }
+            fclose (fd);
+            g_free (scsidel);
+            g_free (fcphbasysfs);
+            g_free (fcpwwpnsysfs);
+            g_free (fcplunsysfs);
+        }
+    }
+    fclose (scsifd);
+    g_free (fcpsysfs);
+    return TRUE;
+}
+
+/*
+ * bd_s390_zfcp_offline:
+ * @devno: zfcp device number
+ * @wwpn: zfcp WWPN (World Wide Port Number)
+ * @lun: zfcp LUN (Logical Unit Number)
+ * @error: (out): place to store error (if any)
+ *
+ * Returns: whether a zfcp device was successfully switched offline
+ */
+gboolean bd_s390_zfcp_offline (const gchar *devno, const gchar *wwpn, const gchar *lun, GError **error) {
+    gboolean failed = FALSE;
+    gint rc = 0;
+    FILE *fd = NULL;
+    glob_t luns;
+
+    gchar *zfcpsysfs = "/sys/bus/ccw/drivers/zfcp";
+    gchar *offline = g_strdup_printf ("%s/%s/online", zfcpsysfs, devno);
+    gchar *unitrm = g_strdup_printf ("%s/%s/%s/unit_remove", zfcpsysfs, devno, wwpn);
+    gchar *pattern = g_strdup_printf ("%s/0x??????????????\?\?/0x????????????????", zfcpsysfs);
+    const gchar *chccwdev[4] = {"/usr/sbin/chccwdev", "-d", devno, NULL};
+
+    failed = bd_s390_zfcp_scsi_offline(devno, wwpn, lun, error);
+    if (failed == 0) {
+        g_set_error (error, BD_S390_ERROR, BD_S390_ERROR_DEVICE,
+                     "Could not correctly delete SCSI device of zFCP %s with WWPN %s, LUN %s\n", devno, wwpn, lun);
+        return FALSE;
+    }
+
+    /* remove lun */
+    fd = fopen (unitrm, "w");
+    if (!fd) {
+        g_set_error (error, BD_S390_ERROR, BD_S390_ERROR_DEVICE,
+                     "Failed to open %s", unitrm);
+        return FALSE;
+    }
+    rc = fputs (lun, fd);
+    if (rc == EOF) {
+        g_set_error (error, BD_S390_ERROR, BD_S390_ERROR_DEVICE,
+                     "Could not remove LUN %s at WWPN %s on zFCP device %s\n", lun, wwpn, devno);
+        g_free (unitrm);
+        return FALSE;
+    }
+    fclose (fd);
+    g_free (unitrm);
+    rc = 0;
+
+    /* gather the luns  */
+    rc = glob (pattern, GLOB_ONLYDIR, NULL, &luns);
+    if (rc == GLOB_ABORTED || rc == GLOB_NOSPACE) {
+        g_set_error (error, BD_S390_ERROR, BD_S390_ERROR_DEVICE,
+                     "An error occurred trying to determine if other LUNs are still associated with WWPN %s\n", wwpn);
+        globfree (&luns);
+    }
+    /* check if we have any matches found; if so, bail */
+    if (luns.gl_pathc > 0) {
+        g_set_error (error, BD_S390_ERROR, BD_S390_ERROR_DEVICE,
+                     "Not setting zFCP device offline since it still has other LUNs");
+        globfree (&luns);
+        return FALSE;
+    }
+    globfree (&luns);
+    rc = 0;
+
+    /* offline */
+    rc = bd_utils_exec_and_report_error (chccwdev, NULL, error);
+    g_free (offline);
+    if (rc == EOF) {
```

Comment:
I'm quite sure the exec function never returns ``EOF`` except for the stupidity of *C* having thousands names for 0.